openedx_certificates/models.py

```diff
@@ -1,5 +1,345 @@
+            )
+            certificate.user_full_name = user_full_name
+            certificate.status = ExternalCertificate.Status.GENERATING
+            certificate.generation_task_id = celery_task_id
+            certificate.save()
+        except ExternalCertificate.DoesNotExist:
+            certificate = ExternalCertificate.objects.create(
+                user_id=user_id,
+                user_full_name=user_full_name,
+                course_id=self.course_id,
+                certificate_type=self.certificate_type.name,
+                status=ExternalCertificate.Status.GENERATING,
+                generation_task_id=celery_task_id,
+            )
+
+        try:
+            generation_module_name, generation_func_name = self.certificate_type.generation_func.rsplit('.', 1)
+            generation_module = import_module(generation_module_name)
+            generation_func = getattr(generation_module, generation_func_name)
+
+            # Run the functions. We do not validate them here, as they are validated in the model's clean() method.
+            certificate.download_url = generation_func(self.course_id, user, certificate.uuid, custom_options)
+            certificate.status = ExternalCertificate.Status.AVAILABLE
+            certificate.save()
+        except Exception as exc:  # noqa: BLE001
+            certificate.status = ExternalCertificate.Status.ERROR
+            certificate.save()
+            msg = f'Failed to generate the {certificate.uuid=} for {user_id=} with {self.id=}.'
+            raise CertificateGenerationError(msg) from exc
+
+
+class ExternalCertificate(TimeStampedModel):
+    """
+    Model to represent each individual certificate awarded to a user for a course.
+
+    This model contains information about the related course, the user who earned the certificate,
+    the download URL for the certificate PDF, and the associated certificate generation task.
+
+    .. note:: The ID field is not a conventional auto-incrementing integer, but a value
+       that allows for old certificates with custom IDs.
+
+    .. pii: The User's name is stored in this model.
+    .. pii_types: id, name
+    .. pii_retirement: retained
+    """
+
+    class Status(models.TextChoices):
+        """Status of the certificate generation task."""
+
+        GENERATING = 'generating', _('Generating')
+        AVAILABLE = 'available', _('Available')
+        ERROR = 'error', _('Error')
+        INVALIDATED = 'invalidated', _('Invalidated')
+
+    uuid = models.UUIDField(
+        primary_key=True,
+        default=uuid.uuid4,
+        editable=False,
+        help_text=_('Auto-generated UUID of the certificate'),
+    )
+    user_id = models.IntegerField(help_text=_('ID of the user receiving the certificate'))
+    user_full_name = models.CharField(max_length=255, help_text=_('User receiving the certificate'))
+    course_id = CourseKeyField(max_length=255, help_text=_('ID of a course for which the certificate was issued'))
```

Comment:
Why do we store the user's full name in a separate field, but we don't store the course name in a separate field?